llama/llama.go

```diff
@@ -0,0 +1,512 @@
+	cparams.main_gpu = C.int32_t(params.MainGpu)
+	cparams.use_mmap = C.bool(params.UseMmap)
+	cparams.use_mlock = C.bool(params.UseMlock)
+	cparams.vocab_only = C.bool(params.VocabOnly)
+
+	if len(params.TensorSplit) > 0 {
+		tensorSplitData := &params.TensorSplit[0]
+
+		var tensorSplitPin runtime.Pinner
+		tensorSplitPin.Pin(tensorSplitData)
+		defer tensorSplitPin.Unpin()
+
+		cparams.tensor_split = (*C.float)(unsafe.Pointer(tensorSplitData))
+	}
+
+	if params.Progress != nil {
+		handle := cgo.NewHandle(params.Progress)
+		defer handle.Delete()
+
+		var handlePin runtime.Pinner
+		handlePin.Pin(&handle)
+		defer handlePin.Unpin()
+
+		cparams.progress_callback = C.llama_progress_callback(C.llamaProgressCallback)
+		cparams.progress_callback_user_data = unsafe.Pointer(&handle)
+	}
+
+	return &Model{c: C.llama_load_model_from_file(C.CString(modelPath), cparams)}
+}
+
+func FreeModel(model *Model) {
+	C.llama_free_model(model.c)
+}
+
+func NewContextWithModel(model *Model, params ContextParams) *Context {
+	return &Context{
+		c:          C.llama_new_context_with_model(model.c, params.c),
+		numThreads: int(params.c.n_threads),
+	}
+}
+
+func (m *Model) NumVocab() int {
+	return int(C.llama_n_vocab(m.c))
+}
+
+func (m *Model) TokenIsEog(token int) bool {
+	return bool(C.llama_token_is_eog(m.c, C.llama_token(token)))
+}
+
+func (m *Model) AddBOSToken() bool {
+	return bool(C.llama_add_bos_token(m.c))
+}
+
+func (m *Model) ApplyLoraFromFile(context *Context, loraPath string, scale float32, threads int) error {
+	cLoraPath := C.CString(loraPath)
+	defer C.free(unsafe.Pointer(cLoraPath))
+
+	loraAdapter := C.llama_lora_adapter_init(m.c, cLoraPath)
+
+	err := -1
+	if loraAdapter != nil {
+		err = int(C.llama_lora_adapter_set(context.c, loraAdapter, C.float(scale)))
+	}
+	if err != 0 {
+		return errors.New("error applying lora from file")
+	}
+
+	return nil
+}
+
+type Batch struct {
+	c         C.struct_llama_batch
+	batchSize int
+	embedSize int
+}
+
+// Creates a new batch for either word tokens if embed is 0 or
+// image embeddings if embed is specified. Batches cannot contain
+// both types at the same time
+func NewBatch(nTokens int, embed int, maxSeq int) *Batch {
+	return &Batch{
+		c:         C.llama_batch_init(C.int(nTokens), C.int(embed), C.int(maxSeq)),
+		batchSize: nTokens,
+		embedSize: embed,
+	}
+}
+
+func (b *Batch) NumTokens() int {
+	return int(b.c.n_tokens)
+}
+
+func (b *Batch) IsEmbedding() bool {
+	return b.embedSize != 0
+}
+
+// Add adds either a token or an image embedding to the batch depending on the type
+// when the batch was initialized. The other argument will be ignored. Adds to the
+// batch with the given position for the given sequence ids, and optionally instructs
+// to include logits.
+func (b *Batch) Add(token int, embed []float32, pos int, seqIds []int, logits bool) {
+	if !b.IsEmbedding() {
+		unsafe.Slice(b.c.token, b.batchSize)[b.c.n_tokens] = C.llama_token(token)
```

Comment:
Panicing here (and same for the other locations mentioned) will take down the entire runner if there is a corner case with a single completion. While in most cases we will probably hit an assertion in llama.cpp anyways, I would rather try to ride it out.

I've seen a number of cases where asserts have taken down production systems in situations where they otherwise would have survived.